### DISCUSSION TITLE
Call PL/SQL Without Direct usage of package

### DISCUSSION THREAD

**Discussion by hmmftg** [sijms/go-ora#120] [Q&A] [answered]
Hi 
I managed to call plsql package with direct package usage
as you said earlier, to use connection pool, we have to use sql package
can you please help it with this working code:

	import (
		"strings"

		go_ora "github.com/sijms/go-ora/v2"
	)

	type WsModel struct {
		//DB *sql.DB
		DB *go_ora.Connection
	}

	func (m *WsModel) CallPlsql(json string) (int, string, error) {
		var retDb int
		var sb strings.Builder
		sb.WriteString("BEGIN\n\t:1 := SCHEMA.PACKAGE.FUNCTION(:2);\nEND;")
		//sb.WriteString(authJson)
		//sb.WriteString("');\nEND;")
		finalStmnt := sb.String()
		println(finalStmnt)
		stmt := go_ora.NewStmt(finalStmnt, m.DB)
		defer stmt.Close()

		stmt.AddParam("1", retDb, 3, go_ora.Output)
		stmt.AddParam("2", json , len(json), go_ora.Input)

		_, err := stmt.Exec(nil)
		if err != nil {
			return -3, "execute", err
		}

		return retDb, "empty", nil
	}

**Comment by sijms** [comment]
To use connection pool you need to use go sql interface 
```
conn, err := sql.Open("oracle", databaseURL)
if err != nil {
	fmt.Println(err)
	return
}
conn.SetMaxOpenConns(5)
```
when you call conn.Query or conn.Exec 2nd time before close the first connection it will open a new connection and so on
when you close the connection it will return to the pool

**Reply by hmmftg** [reply]
Thanks, but my question is how to call oracle Stored Program
I tried sq.exec but got invalid network data type

**Reply by hmmftg** [reply]
the connection is made using sql.Open method

**Reply by sijms** [reply]
the problem you have output parameter and go sql not support output parameter

**Reply by hmmftg** [reply]
sql.qo line 418: 
`// Out may be used to retrieve OUTPUT value parameters from stored procedures.`
sql.go line 426:
`type Out struct {`

**Comment by sijms** [comment]
if this is the problem you can implement a simple array of connections and use sync.mutex when read/write to this array to avoid race condition

**Comment by sijms** [comment]
I will try to find a solution for this in the future to make a link between golang sql interface and my package

**Reply by hmmftg** [reply]
trying sql.Exec with OUT paramater, gives this error ORA-03115: unsupported network datatype or representation

**Reply by hmmftg** [reply]
`var outParam string`
`_, err = db1.Exec("SP", "123", sql.Out{Dest: &outParam})`

am I doing something wrong?

**Reply by sijms** [reply]
I am working on it now

**Reply by sijms** [reply]
I add support for it you can test now

**Reply by sijms** [reply]
the problem now you need to specify the size of string thus you need to fill string with appropriate size using
`strings.Repeat(" ", 60)` or whatever size you want before pass it to the sql.Out

**Comment by sijms** [comment, marked as answer]
I make an example called output_pars

**Reply by hmmftg** [reply]
Thanks a lot 
your package is awesome, no gccgo and very hardwork to remove buggy oci 
We replace other drivers with your pure go driver, with love ;)

**Reply by sijms** [reply]
you are welcome
